metricproviders/influxdb/influxdb.go

```diff
@@ -0,0 +1,136 @@
+	ProviderType = "Influxdb"
+	//DefaultInfluxdbTokensSecretName is the k8s secret that has InfluxDB api token, org and address
+	DefaultInfluxdbTokensSecretName = "influxdb"
+	influxdbToken                   = "authToken"
+	influxdbOrg                     = "org"
+	influxdbAddress                 = "address"
+	defaultQueryTimeout             = 30 * time.Second
+)
+
+// Provider contains all the required components to run a influxdb flux query
+type Provider struct {
+	api    influxapi.QueryAPI
+	logCtx log.Entry
+}
+
+// Type indicates provider is a influxdb provider
+func (p *Provider) Type() string {
+	return ProviderType
+}
+
+// GetMetadata returns any additional metadata which needs to be stored & displayed as part of the metrics result.
+func (p *Provider) GetMetadata(metric v1alpha1.Metric) map[string]string {
+	return nil
+}
+
+// Run queries influxdb for the metric
+func (p *Provider) Run(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric) v1alpha1.Measurement {
+	startTime := metav1.Now()
+	newMeasurement := v1alpha1.Measurement{
+		StartedAt: &startTime,
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), defaultQueryTimeout)
+	defer cancel()
+	result, err := p.api.Query(ctx, metric.Provider.Influxdb.Query)
+	if err != nil {
+		return metricutil.MarkMeasurementError(newMeasurement, err)
+	}
+	newValue, newStatus, err := p.processResponse(metric, result)
+	if err != nil {
+		return metricutil.MarkMeasurementError(newMeasurement, err)
+	}
+	newMeasurement.Value = newValue
+
+	newMeasurement.Phase = newStatus
+	finishedTime := metav1.Now()
+	newMeasurement.FinishedAt = &finishedTime
+	return newMeasurement
+}
+
+// Resume should not be used by the influxdb provider since all the work should occur in the Run method.
+func (p *Provider) Resume(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric, measurement v1alpha1.Measurement) v1alpha1.Measurement {
+	p.logCtx.Warn("Influxdb provider should not execute the Resume method")
+	return measurement
+}
+
+// Terminate should not be used by the influxdb provider since all the work should occur in the Run method.
+func (p *Provider) Terminate(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric, measurement v1alpha1.Measurement) v1alpha1.Measurement {
+	p.logCtx.Warn("Influxdb provider should not execute the Terminate method")
+	return measurement
+}
+
+// GarbageCollect is a no-op for the influxdb provider
+func (p *Provider) GarbageCollect(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric, limit int) error {
+	return nil
+}
+
+func (p *Provider) processResponse(metric v1alpha1.Metric, result *influxapi.QueryTableResult) (string, v1alpha1.AnalysisPhase, error) {
+	var res []interface{}
+	if result == nil {
+		return "", v1alpha1.AnalysisPhaseError, fmt.Errorf("no results returned from flux query")
+	}
+	for result.Next() {
+		res = append(res, result.Record().Value())
+	}
+	if len(res) == 0 {
+		return "", v1alpha1.AnalysisPhaseError, fmt.Errorf("no results returned from flux query")
```

Comment:
Shall we distinguish the error message from `result==nil` (line 91)?

Reply:
Done here 878fe31196303ca9fe97f4e7b5321d7c4cade27b